File: src/Javascript/array/75.颜色分类.js
/*
 * @lc app=leetcode.cn id=75 lang=javascript
 *
 * [75] 颜色分类
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    let p0 = 0, p2 = nums.length - 1;
    for (let i = 0; i <= p2; i++) {
        while (i <= p2 && nums[i] === 2) swap(nums, i, p2--);
        if (nums[i] === 0) swap(nums, i, p0++);
    }
};
const swap = (nums, i, j) => {
    const tmp = nums[i];
    nums[i] = nums[j];
    nums[j] = tmp;
}
// @lc code=end

File: src/Java/Mirosoft/75.颜色分类.java
/*
 * @lc app=leetcode.cn id=75 lang=java
 *
 * [75] 颜色分类
 */

// @lc code=start
class Solution {
    public void sortColors(int[] nums) {
        int lo = 0, hi = nums.length - 1;
        for (int i = 0; i < nums.length; i++) {
            while (i < hi && nums[i] == 2) swap(nums, hi--, i);
            if (nums[i] == 0) swap(nums, lo++, i);
        }
    }
    public void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}
// @lc code=end

